### DISCUSSION TITLE
Usage of expressions and contexts in `uses` clause

### DISCUSSION THREAD

**Discussion by kellervater** [community/community#27048] [Actions]
<p>I’ve got a use case where I want to test my action in an example workflow.<br>
This works out great for me with hardcoded values in the <code>uses</code> clause. But it would be much more elegant to be able to use context variables like <code>github.repository</code> in the <code>uses</code> clause to be fork agnostic.</p>
<p>Something like this:</p>
<pre><code>...
        uses: "${{ github.repository }}@master"
...
</code></pre>
<p>In my case I test a fork of a Github Repo. This means the root of my fork has another repository name than I do.<br>
So it’ll be great if this repository could be resolved dynamically by using the github context.</p>
<p>This is currently not possible and ends with:</p>
<pre><code>### ERRORED 11:55:40Z

- Your workflow file was invalid: The pipeline is not valid. .github/workflows/gpr.yaml (Line: 14, Col: 13): Unrecognized named-value: 'github'. Located at position 1 within expression: github.repository
</code></pre>

**Comment by wipe2238** [comment, marked as answer]
<p>You can use <a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#example-using-action-in-the-same-repository-as-the-workflow" rel="nofollow noopener">path</a> as argument for <code>uses</code>, so this should do the trick</p>
<pre><code>- uses: ./
</code></pre>

**Comment by kellervater** [comment]
<p>Thank you very much. This did the trick. Kudos to you :smiley:</p>

**Comment by chrispat** [comment]
<p>For a bit more color.  Because we have the settings where a repo or org owner can specify if they want third party actions or not we need to be able to resolve the full list of referenced actions before we start the workflow.  If we enabled expressions in the uses section that would not be possible until runtime and we would potentially get hours into a job before it failed due to this policy.</p>

**Reply by jenseng** [reply]
> we need to be able to resolve the full list of referenced actions before we start the workflow

That's not quite accurate, as only the top-level actions are resolved; if the job `uses` a composite action, that job's own `uses` are not resolved until the action runs. Given that, I wonder if it would be worth revisiting this top-level behavior, as it [makes it difficult](https://github.com/actions/runner/issues/895#issuecomment-1403911529) to build and test dependent public actions within a monorepo.

E.g. If you're developing multiple actions in a monorepo, ideally you might want to do something like this:

**file: SameOrg/same_repo/some_action/action.yml**
```yaml
# ensure we have the same sha/ref for `some_action` and `another_action`, whether
# we're using this externally (e.g. `uses: SameOrg/same_repo/some_action@v1`)
# or internally (e.g. PR/CICD workflow that does `uses: ./some_action`)
uses: SameOrg/same_repo/another_action@${{ github.repo == 'SameOrg/same_repo' && github.sha || github.action_ref }}
```

**Reply by krebemm** [reply]
how would you use this expression with both dynamic and static variables?

**Reply by jenseng** [reply]
By static do you mean a hardcoded value? Within expressions, you can use literal strings. For example, this expression hardcodes the `v1` tag if we're in another repo:

```yaml
${{ github.repo == 'SameOrg/same_repo' && github.sha || 'v1' }}
```

**Reply by krebemm** [reply]
yes, hard coded. i was thinking something like:

${{ github.repo == 'SameOrg/same_repo' && 'static value here'github.sha || 'v1' }}

but i actually just found a fix by just adding it outside of the expression like
'static value here' ${{ github.repo == 'SameOrg/same_repo' && github.sha || 'v1' }}

**Reply by jenseng** [reply]
When I need to mix dynamic and static stuff in expressions, another useful thing i've found is the [`format` function](https://docs.github.com/en/actions/learn-github-actions/expressions#format), e.g.

```yaml
${{ github.repo == 'SameOrg/same_repo' && format('static value here{0}', github.sha) || 'v1' }}
```

**Reply by krebemm** [reply]
awesome! thank you for the syntax

**Comment by kellervater** [comment]
<p>Thank you for this explanation</p>

**Comment by eps1lon** [comment]
<p>Do you have a working example for this? When I tried this it was not doing anything: <a href="https://github.com/eps1lon/actions-label-merge-conflict/blob/2f617ca50dfb880adbf628a61e3d4f427561a1df/.github/workflows/ci.yml#L24" rel="nofollow noopener">https://github.com/eps1lon/actions-label-merge-conflict/blob/2f617ca50dfb880adbf628a61e3d4f427561a1df/.github/workflows/ci.yml#L24</a></p>

**Comment by atoomic** [comment]
<p>This works fine for me if you have an action.yml in your root directory</p>
<pre><code>- uses: actions/checkout@v2
- uses: ./
</code></pre>

**Comment by mktcode** [comment]
<p><a class="mention" href="/wipe2238">@wipe2238</a> The link to the docs seems to be outdated.</p>
<p><a href="https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#example-using-an-action-in-the-same-repository-as-the-workflow" class="onebox" target="_blank" rel="noopener nofollow ugc">https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#example-using-an-action-in-the-same-repository-as-the-workflow</a></p>

**Comment by jenseng** [comment]
You can work around this with nested composite actions; although the job step's `uses` need to be static/resolvable, if one of those actions is a composite action, its own `uses` won't be resolved until the action runs. So we can use this to our advantage to dynamically build a composite action that runs our desired action. It's a tad convoluted, so I made a [generic action](https://github.com/marketplace/actions/dynamic-uses) to make this a little more sane 😅 

Given a step like so:

```yaml
- uses: actions/setup-node@v3
  with:
    node-version: 18
```

If you want your `uses` to be dynamic you can do:

```yaml
- uses: jenseng/dynamic-uses@v1
  with:
    # now you can use expressions 🥳
    uses: actions/setup-node@${{ inputs.version }}
    # the `with` needs to be converted to a valid json string
    with: '{ "node-version": 18 }'
```

**Comment by timeasch** [comment]
Z

**Comment by timeasch** [comment]
Oo

**Reply by timeasch** [reply]
Nl

**Comment by parasavellipmg** [comment]
Any idea why I couldn't use expression in "uses" ? Like

- uses: actions/setup-node@${{ vars.SETUP_NODE_VERSION }}
  with:
    node-version: 18
